internal/highlight/highlight.go

```
@@ -450,9 +456,27 @@ func unhighlightLongLines(h string, n int) (string, error) {
 	return buf.String(), nil
 }
 
-// ParseLinesFromHighlight takes the highlighted html table and returns a slice of highlighted strings, where each string corresponds a single line in the original, highlighed file.
-func ParseLinesFromHighlight(input string) ([]string, error) {
-	doc, err := html.Parse(strings.NewReader(input))
+// Lines highlights the file and returns a list of highlighted lines.
+// The returned boolean represents whether or not highlighting was aborted due
+// to timeout.
+func Lines(ctx context.Context, p Params) ([]string, bool, error) {
+
+	html, aborted, err := Code(ctx, p)
+	if err != nil {
+		if err == errBinary {
+			return nil, aborted, nil
```

This means if the input was binary and could not be highlighted at all, you return no HTML *and* no error. That seems a bad API to me. I think you should return the error and handle it properly at the call sites instead.

---

Additionally, get rid of `errBinary` once you do this because it becomes useless again.

---

It was meant to address your concern in the issue:

> Binary-file checks like this, this, and this appear to be sprinkled throughout the code. This smells to me of defensive coding, especially because highlight.Code already handles binary files appropriately. These checks should not be needed anywhere I believe, because if the file is binary highlight.Code is going to give an error which should be handled without a prior "is binary" check.

We _do_ allow binary renders to fail, as we still might have a non-binary head file for example, and hence it would be just additions of the head in the final output.
I think returning `[]string{}` should be good here instead to make that intention more clear? Then, we wouldn't have to expose the Error so it can be checked on in referencing code. Otherwise, we can get rid of it, but would have to keep the error around, for assertions

---

The concern of mine you quoted was primarily about the fact that we checked if the file was binary _before_ passing it to `highlight.Foobar`, instead of relying on `highlight.Foobar` to detect this and tell us it is binary. Also, it was about the fact that we did this in several places.

As I understand it now, that problem is mostly addressed and you can just expose `ErrBinary` now and check it once here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@es/improve-filediffs/-/blob/cmd/frontend/graphqlbackend/repository_comparison.go#L482-488

I think that is OK and desirable because, logically, if you try to highlight a binary file for display you would expect to get an error and not simply "no result" (empty lines).

In other words, empty lines make sense for your use case of `highlight.CodeAsLines` but not other potential use cases, and thus we should keep that logic close to your use case. Do you buy that argument?

---

absolutely! adjusting it now